cdragontoolbox/tftdata.py

```diff
@@ -65,17 +67,70 @@ def replace_in_data(data):
                     if key in data and data[key] in replacements:
                         data[key] = replacements[data[key]]
 
-            for data in template["champs"]:
+            for data in [template["sets"][x]["traits"] for x in template["sets"]]:
+                replace_in_data(data)
+            for data in [template["sets"][x]["champions"] for x in template["sets"]]:
                 replace_in_data(data)
                 if "ability" in data:
                     replace_in_data(data["ability"])
             for data in template["items"]:
                 replace_in_data(data)
-            for data in template["traits"]:
-                replace_in_data(data)
 
             with open(os.path.join(output, f"{lang}.json"), "w", encoding="utf-8") as f:
-                json.dump(template, f, cls=NaiveJsonEncoder, indent=4, sort_keys=True)
+                json.dump(template, f, cls=NaiveJsonEncoder, indent=4)
+
+    def generate_correct_set_info(self, sets, traits, champs):
+
+        set_map = {}
+        for key, value in sets.items():
+            current_set = {"name": value["name"]}
+
+            set_champs = []
+            for char_name in value["internal"]["characters"]:
+                for real_char in champs:
+                    if char_name == real_char["internal"]["name"]:
+                        set_champs.append(real_char)
```

Comment:
As for previous PR, you don't need those `internal` entries. `champs` is only used for its "internal names" and `sets` is only used for its "internal characters", so no need to mix them with other values.

Furthermore, this will ease computations made in this function.

Reply:
I'm not sure what change you would propose here?

Reply:
Instead of having `sets = {key: {"internal": internal_data, "other": ...}}` you can have `sets = {key: internal_data}` directly.
You should be able to do something similar for the other `"internal"` (the one in`traits`).

Reply:
In both of the cases, the parent object pass along other data. so i dont see that as possible?

Reply:
The non-internal data does not seem to be used, so you don't need to merge internal and non-internal data.

Reply:
in both cases the `internal` data is used as a lookup object to merge the full data into the set, so it is used.

Reply:
But the full data is not from `sets` is not merged (only `"name"` is used, I missed this one at first).
The data that is merged comes from `champs`, for which you also use an `"internal"` (which would be a more difficult to remove, I agree).

With the current design, code flow (more accurately, the data flow) is difficult to follow. It abuses mutability of data sets that should be immutable. The method return a built value but also modifies its parameters as a side effect.

Design-wise, I don't see a good reason to put `sets[key]["internal"]` into `sets[key]`; both are of different nature and should not be mixed.
You can either store pairs of internal/non-internal data (something like `sets[key] = (not_internal, internal)`)`, or use separate variables with the same indexing. And that may allow you to use a different indexing for the internal data, in order to avoid some of the "real char name" lookups.

Reply:
imo its the same if you store 
`sets[x][internal] = a 
sets[x][real] = b` 
vs 
`sets[x] = (internal, real)`

Reply:
I agree, but in your case you use neither.
You have `sets[x] = b` and `b["internal"] = a`

And there is also the fact that you update the parameters, which IMHO is the biggest issue. Having the `"internal"` field wouldn't be an issue if it was just considered as part of the data instead of being added temporarily before being hacked away.